### ISSUE TITLE
Frontend crash in get_rest_url

### ISSUE THREAD

**Issue by roborourke** [humanmade/Mercator-Redirect#1]
Original issue: https://github.com/humanmade/Mercator/issues/86

> After installing Mercator, the frontend of one of my sites (haven't checkt the others) crashes with the following stack trace:
>
> `Fatal error:  Uncaught Error: Call to a member function using_index_permalinks() on null in /path/wordpressmu/wp-includes/rest-api.php:314
> Stack trace:
> ```
> #0 /path/wordpressmu/wp-includes/rest-api.php(363): get_rest_url(NULL, '/', 'json')
> #1 /path/wordpressmu/wp-content/mu-plugins/mercator-redirect/redirect.php(64): rest_url()
> #2 /path/wordpressmu/wp-includes/class-wp-hook.php(298): Mercator\Redirect\handle_redirect('')
> #3 /path/wordpressmu/wp-includes/class-wp-hook.php(323): WP_Hook->apply_filters(NULL, Array)
> #4 /path/wordpressmu/wp-includes/plugin.php(453): WP_Hook->do_action(Array)
> #5 /path/wordpressmu/wp-settings.php(326): do_action('plugins_loaded')
> #6 /path/wordpressmu/wp-config.php(226): require_once('/www/vhosts/wor...')
> #7 /path/wordpressmu/wp-load.php(37): require_once('/www/vhosts/wor...')
> #8 /path/wordpressmu/wp-blog-header.php(13): require_once('/www/vhosts/wor...')
> #9 /path/wordpressmu/index.php(17): require('/www/vhosts/wor...')
> #10 {main}
>  thr in /path/wordpressmu/wp-includes/rest-api.php on line 314`
> ```
>
> It looks like `global $wp_rewrite;` is null at that point.

**Comment by roborourke** [comment]
@mundschenk-at can you paste the contents of your `sunrise.php`?

**Comment by mundschenk-at** [comment]
I can't really say if that was the version I used back then (I've since gone back to the standard Domain Mapping plugin), but that's what I've got lying around on disk:

```
<?php
// Default mu-plugins directory if you haven't set it
defined( 'WPMU_PLUGIN_DIR' ) or define( 'WPMU_PLUGIN_DIR', WP_CONTENT_DIR . '/mu-plugins' );

require WPMU_PLUGIN_DIR . '/mercator/mercator.php';
require WPMU_PLUGIN_DIR . '/mercator-redirect/redirect.php';
```

**Comment by ben-morin** [comment]
I'm getting this too... has to be because the global $wp_rewrite is null at the time rest_url() is called in handle_redirect()

wp-includes/rest-api.php:
```
function get_rest_url( $blog_id = null, $path = '/', $scheme = 'rest' ) {
    if ( empty( $path ) ) {
        $path = '/';
    }
 
    if ( is_multisite() && get_blog_option( $blog_id, 'permalink_structure' ) || get_option( 'permalink_structure' ) ) {
        global $wp_rewrite;

       if ( $wp_rewrite->using_index_permalinks() ) {
            $url = get_home_url( $blog_id, $wp_rewrite->index . '/' . rest_get_url_prefix(), $scheme );
        } else {
            $url = get_home_url( $blog_id, rest_get_url_prefix(), $scheme );
        }

```

**Comment by ben-morin** [comment]
It's defined in wp-includes/rewrite.php as a single instance global variable, $wp_rewrite, is initialized in wp-settings.php.  

wp-settings.php is loaded in wp-config.php and that should happen before sunrise.php ever runs, right?  So I'm not sure why $wp_rewrite would ever be null.

**Comment by ben-morin** [comment]
Well... ```do_action( 'plugins_loaded' );``` happens before ```$GLOBALS['wp_rewrite'] = new WP_Rewrite();``` so that's interesting.

Probably going to have to add handle_redirect to the setup_theme action instead of plugins_loaded.

**Comment by roborourke** [comment]
@ben-morin think this will be fixed in #4 - still better to redirect as early as possible. It is a weird issue though. I'll raise it with the API team.

**Comment by roborourke** [comment]
And thanks for digging into it 👍
